### PR TITLE
Added volume list stack.

### PR DESCRIPTION
**Pull request by anastasiamac**
Added volume list.

### CODE REVIEW

File: api/storage/client.go [hunk 1]
```diff
@@ -82,3 +82,15 @@ func (c *Client) CreatePool(pname, ptype string, pconfig map[string]interface{})
 	}
 	return c.facade.FacadeCall("CreatePool", args, nil)
 }
+
+// ListVolumes lists volumes according to a given filter.
+// If no filter was provided, this will return a list
+// of all volumes.
+func (c *Client) ListVolumes(machines []string) ([]params.StorageDisk, error) {
```
Comment (axw) [inline api/storage/client.go]: I'd like to cut down on the terminology, which is why I winced a bit when the suggestion of Disk came up in the standup. We can defer this, but I think we may want to just auto-create Volumes from discovered block devices if there are no matching provider-created Volumes.

File: api/storage/client_test.go [hunk 2]
```diff
@@ -145,3 +145,54 @@ func (s *storageMockSuite) TestCreatePool(c *gc.C) {
...
+			c.Check(request, gc.Equals, "ListVolumes")
+
+			args, ok := a.(params.StorageVolumeFilter)
+			c.Assert(ok, jc.IsTrue)
+			c.Assert(args.Machines, gc.HasLen, 2)
+
+			if results, k := result.(*params.StorageVolumesResult); k {
+				disks := make([]params.StorageDisk, len(args.Machines))
+				for i, m := range args.Machines {
+					attachment := params.StorageAttachment{
```
Comment (axw) [inline api/storage/client_test.go]: Please call this VolumeAttachment. A StorageAttachment is something else: an attachment of a storage instance to a *unit*.

File: api/storage/client_test.go [hunk 3]
```diff
@@ -145,3 +145,54 @@ func (s *storageMockSuite) TestCreatePool(c *gc.C) {
...
+						Tag:         fmt.Sprintf("tag%v", i),
```
Comment (axw) [inline api/storage/client_test.go]: I'm not sure volume attachments will have tags. They're uniquely identified by (storage-ID, machine-ID). Can you please leave this out?

File: api/storage/client_test.go [hunk 4]
```diff
@@ -145,3 +145,54 @@ func (s *storageMockSuite) TestCreatePool(c *gc.C) {
...
+						StorageId:   fmt.Sprintf("storage%v", i),
```
Comment (axw) [inline api/storage/client_test.go]: Can you please make this a (string representation of a) names.StorageTag?

File: api/storage/client_test.go [hunk 5]
```diff
@@ -145,3 +145,54 @@ func (s *storageMockSuite) TestCreatePool(c *gc.C) {
...
+						Assigned:    true,
+						MachineId:   m,
```
Comment (axw) [inline api/storage/client_test.go]: Can you please make this a (string representation of a) names.MachineTag?

File: api/storage/client_test.go [hunk 6]
```diff
@@ -145,3 +145,54 @@ func (s *storageMockSuite) TestCreatePool(c *gc.C) {
...
+			c.Assert(ok, jc.IsTrue)
+			c.Assert(args.Machines, gc.HasLen, 2)
+
+			if results, k := result.(*params.StorageVolumesResult); k {
+				disks := make([]params.StorageDisk, len(args.Machines))
+				for i, m := range args.Machines {
+					attachment := params.StorageAttachment{
+						Tag:         fmt.Sprintf("tag%v", i),
+						StorageId:   fmt.Sprintf("storage%v", i),
+						Assigned:    true,
```
Comment (axw) [inline api/storage/client_test.go]: Assigned is implied by StorageTag != ""

File: api/storage/client_test.go [hunk 7]
```diff
@@ -145,3 +145,54 @@ func (s *storageMockSuite) TestCreatePool(c *gc.C) {
...
+						MachineId:   m,
+						Attached:    true,
+						DeviceName:  fmt.Sprintf("deviceName%v", i),
+						UUID:        fmt.Sprintf("UUIDd%v", i),
```
Comment (axw) [inline api/storage/client_test.go]: I think we should probably drop UUID and Label for now. It's a bit low-level.

File: api/storage/client_test.go [hunk 8]
```diff
@@ -145,3 +145,54 @@ func (s *storageMockSuite) TestCreatePool(c *gc.C) {
...
+						Label:       fmt.Sprintf("label%v", i),
+						Size:        1,
+						InUse:       true,
```
Comment (axw) [inline api/storage/client_test.go]: InUse is for internal use, and shouldn't be exposed.

File: api/storage/client_test.go [hunk 9]
```diff
@@ -145,3 +145,54 @@ func (s *storageMockSuite) TestCreatePool(c *gc.C) {
...
+						FSType:      fmt.Sprintf("fstype%v", i),
```
Comment (axw) [inline api/storage/client_test.go]: Not sure this should be exposed or not. Leave for now, but please spell as Filesystem.

File: apiserver/storage/storage.go [hunk 10]
```diff
@@ -200,3 +201,64 @@ func (a *API) checkProviderTypeSupported(providerType storage.ProviderType) erro
...
+// if disk contains attachments that match filter
+func filterDisk(machineSet set.Strings, disk volume.Disk) (params.StorageDisk, bool) {
+	attachments := []params.StorageAttachment{}
+	for _, attachment := range disk.Attachments() {
+		if one, k := filterAttachment(machineSet, attachment); k {
+			attachments = append(attachments, one)
+		}
+	}
+	// it's possible that there will be no attachments on this disk
+	// that match the filter. This disk will be filtered out too then :D
```
Comment (axw) [inline apiserver/storage/storage.go]: are you saying that we won't ever display unattached volumes? we do want to see them.

File: cmd/juju/storage/volumelist.go [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+	for _, disk := range disks {
+		for _, attachment := range disk.Attachments {
+			fmt.Fprintf(tw, "%s\t%t\t%s\t%s\t%s\t%s\t%d\n",
+				attachment.Tag,
+				attachment.Attached,
+				attachment.MachineId,
+				attachment.DeviceName,
+				attachment.UUID,
+				attachment.Label,
+				attachment.Size,
```
Comment (axw) [inline cmd/juju/storage/volumelist.go]: Please use go-humanize for the size (see my changes to "storage list"). I think we should only do this for the tabular format; the other formats are more likely to be used programmatically.

File: cmd/juju/storage/volumelist.go [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+		Purpose: "list storage volumes",
+		Doc:     VolumeListCommandDoc,
+	}
+}
+
+// SetFlags implements Command.SetFlags.
+func (c *VolumeListCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.StorageCommandBase.SetFlags(f)
+
+	c.out.AddFlags(f, "yaml", map[string]cmd.Formatter{
```
Comment (axw) [inline cmd/juju/storage/volumelist.go]: default to tabular

File: cmd/juju/storage/volumelist_test.go [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+			`"file-system-type":"fstype","provisioned":true}]}`+
+			"]\n",
+	)
+}
+
+func (s *VolumeListSuite) TestVolumeListTabular(c *gc.C) {
+	s.assertValidList(
+		c,
+		[]string{"a", "--format", "tabular"},
+		"TAG  ATTACHED  MACHINE  DEVICE NAME  UUID      LABEL      SIZE\n"+
```
Comment (axw) [inline cmd/juju/storage/volumelist_test.go]: consider using `` for multi-line string constants, would make it easier to determine if the output looks good or not.

File: storage/volume/interface.go [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+// Disk
+type Disk interface {
+	// TODO(anastasiamac 2015-01-30 add name and persisted parameters
+	// when model is decided
+	Attachments() []Attachment
+}
+
+// Attachment is a block device
+type Attachment interface {
```
Comment (axw) [inline storage/volume/interface.go]: FWIW, a lot of that information is specific to the volume, not the attachment. e.g. UUID, Label, Size, FilesystemType, Storage, Provisioned. I'm not too fussed, since this will be replaced soon by structures in the top-level storage package.

File: storage/volume/volume.go [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+	if err != nil {
+		return nil, errors.Annotate(err, "listing block devices")
+	}
+	attachments := make([]Attachment, len(devices))
+	for i, d := range devices {
+		attachments[i] = pm.constructAttachment(d)
+	}
+	// TODO(anastasiamac 2015-01-30) since volumes don't really exist yet,
+	// only return one disk for now
+	volumes := make([]Disk, 1)
```
Comment (axw) [inline storage/volume/volume.go]: Currently, there is a one-to-one relation between BlockDevice (which doubles as VolumeAttachment) and Volume.

File: api/storage/client.go (hunk 1 above)
Comment (anastasiamac) [inline api/storage/client.go]: Added a todo

File: api/storage/client_test.go (hunk 2 above)
Comment (anastasiamac) [inline api/storage/client_test.go]: done

File: api/storage/client_test.go (hunk 3 above)
Comment (anastasiamac) [inline api/storage/client_test.go]: done

File: api/storage/client_test.go (hunk 4 above)
Comment (anastasiamac) [inline api/storage/client_test.go]: done

File: api/storage/client_test.go (hunk 5 above)
Comment (anastasiamac) [inline api/storage/client_test.go]: done.

File: api/storage/client_test.go (hunk 6 above)
Comment (anastasiamac) [inline api/storage/client_test.go]: so it's fine as is. especially since this is test :D 
full stack tests pass as well

File: api/storage/client_test.go (hunk 7 above)
Comment (anastasiamac) [inline api/storage/client_test.go]: done

File: api/storage/client_test.go (hunk 8 above)
Comment (anastasiamac) [inline api/storage/client_test.go]: done

File: api/storage/client_test.go (hunk 9 above)
Comment (anastasiamac) [inline api/storage/client_test.go]: Renamed to FileSystem

File: apiserver/storage/storage.go (hunk 10 above)
Comment (anastasiamac) [inline apiserver/storage/storage.go]: added a todo comment. as per irc, would be best to add this functionality when we can test for it.

File: cmd/juju/storage/volumelist.go (hunk 11 above)
Comment (anastasiamac) [inline cmd/juju/storage/volumelist.go]: will do in another PR with the rest of formatting :D

File: cmd/juju/storage/volumelist.go (hunk 12 above)
Comment (anastasiamac) [inline cmd/juju/storage/volumelist.go]: done

File: cmd/juju/storage/volumelist_test.go (hunk 13 above)
Comment (anastasiamac) [inline cmd/juju/storage/volumelist_test.go]: I've done it in some places but in some instances I find it more confusing them helpful.

File: storage/volume/volume.go (hunk 15 above)
Comment (anastasiamac) [inline storage/volume/volume.go]: Currently, yes :)
I will change all my "Disks" to "Volumes" and remove my old "Volumes" as per irc discussion

File: apiserver/storage/storage.go [hunk 16]
```diff
@@ -200,3 +201,69 @@ func (a *API) checkProviderTypeSupported(providerType storage.ProviderType) erro
...
+
+func (a *API) ListVolumes(filter params.StorageVolumeFilter) (params.StorageVolumesResult, error) {
+	volumeManager := getVolumeManager(a.storage)
+
+	all, err := volumeManager.List()
+	if err != nil {
+		return params.StorageVolumesResult{}, err
+	}
+
+	if len(all) < 1 {
```
Comment (axw) [inline apiserver/storage/storage.go]: you can't overflow a container size, so len will always be >= 0. it's signed so it can be used in for loops and have less surprising behaviour when you do overflow the index

File: api/storage/client_test.go (hunk 6 above)
Comment (axw) [inline api/storage/client_test.go]: What I mean was we don't need to surface this. We can clean it up later.

**Comment by axw** [comment]
LGTM

File: apiserver/storage/storage.go (hunk 16 above)
Comment (anastasiamac) [inline apiserver/storage/storage.go]: I m checking if no volumes were found. Is there a better way to check if a collection (array/slice) is empty?

Comment (axw) [inline apiserver/storage/storage.go]: `len(all) == 0`

Comment (anastasiamac) [inline apiserver/storage/storage.go]: changed :D thnx for suggestion
